File: 1D Dynamic Programming/646. Maximum Length of Pair Chain/Solution.java
import java.util.*;

class Solution {
    private int dfs(int idx, int[][]pairs, int[] dp){
        int n = pairs.length;
        if (idx >= n){
            return 1;
        }
        if (dp[idx + 1] != -1){
            return dp[idx + 1];
        }
        int res = 0;
        for (int i=idx+1; i<n; i++){
            if (idx == -1 || pairs[idx][1] < pairs[i][0]){
                res = Math.max(res, 1 + dfs(i, pairs, dp));
            }
        }
        return dp[idx + 1] = res;
    }
    public int findLongestChain(int[][] pairs) {
        // N: len(pairs)
        // Time Complexity: O(N LogN + N^2)
        // Space Complexity: O(2N) (with recursion stack)
        // int n = pairs.length;
        // Arrays.sort(pairs, (a, b) -> Integer.compare(a[0], b[0]));
        // int[] dp = new int[n + 1];
        // Arrays.fill(dp , -1);
        // return dfs(-1, pairs, dp);


        // N: len(pairs)
        // Time Complexity: O(N LogN + N^2)
        // Space Complexity: O(N) (no recursion stack)
        // int n = pairs.length;
        // Arrays.sort(pairs, (a, b) -> Integer.compare(a[0], b[0]));
        // int[] dp = new int[n + 1];
        // for (int idx=n-1; idx>=-1; idx--){
        //     if (idx >= n) dp[idx + 1] = 1;
        //     int res = 0;
        //     for (int i=idx+1; i<n; i++){
        //         if (idx == -1 || pairs[idx][1] < pairs[i][0]){
        //             res = Math.max(res, 1 + dp[i + 1]);
        //         }
        //     }
        //     dp[idx + 1] = res;
        // }
        // return dp[0];


        // N: len(pairs)
        // Time Complexity: O(N LogN + N)
        // Space Complexity: O(1)
        int n = pairs.length;
        Arrays.sort(pairs, (a, b)->Integer.compare(a[1], b[1]));
        int longestPair = 1;
        int[] prev = pairs[0];
        for (int i=1; i<n; i++){
            if (prev[1] < pairs[i][0]){
                longestPair++;
                prev = pairs[i];
            }
        }
        return longestPair;
    }
}

File: 1D Dynamic Programming/646. Maximum Length of Pair Chain/Solution.py
from typing import List

class Solution:
    def findLongestChain(self, pairs: List[List[int]]) -> int:
        # N: len(pairs)
        # Time Complexity: O(N LogN + N^2)
        # Space Complexity: O(2N) (with recursion stack)
        # n = len(pairs)
        # pairs.sort(key = lambda x: x[0])
        # dp = [-1] * (n + 1)
        # def dfs(idx):
        #     if idx >= n:
        #         return 1
        #     if dp[idx] != -1:
        #         return dp[idx]
        #     res = 0
        #     for i in range(idx + 1, n):
        #         if pairs[idx][1] < pairs[i][0] or idx == -1:
        #             res = max(1 + dfs(i), res)
        #     dp[idx] = res
        #     return dp[idx]
        # return dfs(-1)


        # N: len(pairs)
        # Time Complexity: O(N LogN + N^2)
        # Space Complexity: O(N) (no recursion stack)
        # n = len(pairs)
        # pairs.sort(key = lambda x: x[0])
        # dp = [0] * (n + 1)
        
        # for idx in range(n, -2, -1):
        #     if idx >= n:
        #         dp[idx] = 1
        #     res = 0
        #     for i in range(idx + 1, n):
        #         if pairs[idx][1] < pairs[i][0] or idx == -1:
        #             res = max(1 + dp[i], res)
        #     dp[idx] = res
        # return dp[-1]

        # N: len(pairs)
        # Time Complexity: O(N LogN + N)
        # Space Complexity: O(1)
        n = len(pairs)
        pairs.sort(key = lambda x: x[1])
        longestChain = 1
        prev = pairs[0]
        for i in range(1, n):
            if pairs[i][0] > prev[1]:
                longestChain += 1
                prev = pairs[i]
        return longestChain